### PR TITLE
Endpoint stats interface

### PR DESCRIPTION
**Pull request by ryleung-solana**
In Anza, we need to get more insight into the resources being used by quinn in servicing quic handshakes. This PR seeks to add an initial interface to the Endpoint struct to allow for getting more detailed stats than just the open connections, and some initial data relevant to the handshakes. On a side note to the Quinn devs, if there is a good way to support getting the current number of open handshakes, we would be interested in know it. Currently, it seems that the Endpoint is not signaled when a Connecting future finishes, and to add such an event would require some refactoring, not to mention handling an extra event, locking, etc.

### CODE REVIEW

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:50]: To avoid risk of overflow on long-running 32-bit applications and for consistency with `ConnectionStats`, counters should be `u64`.

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:55]: `_since_last_sample` counters are redundant: a caller interested in progression over time can store the previous total and compute changes. This both simpler for us to implement and less lossy, because a caller can aggregate over any time interval. Similar logic probably applies to `open_connections` too.

Comment (Ralith) [inline quinn/src/endpoint.rs:55]: You probably want these fields to be public.

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:50]:
```suggestion
#[derive(Debug, Default, Copy, Clone)]
pub struct EndpointStats {
```

**Review by Ralith — CHANGES_REQUESTED** [review]
> if there is a good way to support getting the current number of open handshakes, we would be interested in know it

This isn't readily available, because endpoint behavior isn't sensitive to connection state once the application has decided to accept it. What *is* readily available is the number of incoming connection attempts which haven't yet been accepted/rejected/ignored. You could always maintain your own atomic counter that you update by whatever logic you like, of course.

File: quinn/src/endpoint.rs
Comment (djc) [inline quinn/src/endpoint.rs:49]: This should be defined further down, probably between `impl Endpoint` and `struct EndpointDriver(..)`.

**Comment by ryleung-solana** [comment]
@Ralith @djc one thing that we're particularly interested in is related to the resource usage from crypto, especially in accept() because we don't currently reject any incoming connection requests. We merely accept all incoming connections and then immediately drop some Connecting future based on our own logic.

**Comment by djc** [comment]
> @Ralith @djc one thing that we're particularly interested in is related to the resource usage from crypto, especially in accept() because we don't currently reject any incoming connection requests.

Why not? I think we had some Solana people look at that as an option for this exact reason.

**Comment by ryleung-solana** [comment]
> > @Ralith @djc one thing that we're particularly interested in is related to the resource usage from crypto, especially in accept() because we don't currently reject any incoming connection requests.
> 
> Why not? I think we had some Solana people look at that as an option for this exact reason.

Mostly because we currently only do some basic rate-limiting and nothing more. It's probably a good idea to look at using reject when over the rate limit instead of accepting and immediately dropping :sweat: but the thing is that we currently don't really have a great idea of what the limit should be, since we don't have insight into the resource usage from crypto (but we think it's significant).

**Comment by ryleung-solana** [comment]
Would you guys have any ideas of how we might be able to get some better numbers on e.g. the mem usage from crypto in accept() before the future is returned?

**Comment by Ralith** [comment]
Can you elaborate on "mem usage from crypto"? I don't expect cryptographic operations to consume significant memory, or really anything but O(input). Usually the concern with handshakes is CPU usage, which could in principle be measured with some sort of high resolution timer around `accept` and, most importantly, the `poll` calls on the internal connection driver future.

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:328]: Would it be more useful/consistent to lean into the monotonic counter paradigm and count the lifetime number of successfully established and lost connections, rather than reporting the instantaneous live count? I have limited experience with large-scale infrastructure monitoring so maybe I'm getting carried away.

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:329]: Should we also count total accepted/rejected/ignored handshakes?

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:331]: nit: So far 3/4 of these fields have "total" in the name. Is that adding any clarity? Maybe something like "handshakes_{received, sent}" and "incoming_buffer_bytes"?

File: quinn/src/endpoint.rs
Comment (ryleung-solana) [reply quinn/src/endpoint.rs:328]: I think we would find it more useful to just sample the number of open connection at that moment.

Comment (Ralith) [reply quinn/src/endpoint.rs:332]: Would that be adequately addressed by using the existing [`Endpoint::open_connections`](https://docs.rs/quinn-proto/latest/quinn_proto/struct.Endpoint.html#method.open_connections) rather than committing to it as part of `EndpointStats`?

File: quinn/src/endpoint.rs
Comment (Ralith) [inline quinn/src/endpoint.rs:331]:
```suggestion
#[non_exhaustive]
pub struct EndpointStats {
```

**Comment by ryleung-solana** [comment]
> Can you elaborate on "mem usage from crypto"? I don't expect cryptographic operations to consume significant memory, or really anything but O(input). Usually the concern with handshakes is CPU usage, which could in principle be measured with some sort of high resolution timer around `accept` and, most importantly, the `poll` calls on the internal connection driver future.

I think that's (CPU usage) also something to look at as well. We're looking specifically at x509 verification, as it happens before we get the future returned to us when we accept() making it a prime target for rate limiting given some data. It would also allow us to determine once and for all, whether crypto does indeed take a significant amount of memory under attack (e.g. getting spammed with lots of handshake attempts, or some variation of this), which we suspect from a recent security advisory, but don't know for sure.

**Comment by Ralith** [comment]
> We're looking specifically at x509 verification, as it happens before we get the future returned to us when we accept() making it a prime target for rate limiting given some data.

This is no longer the case as of 0.11. No CPU-intensive work is performed until you accept a connection.

**Comment by ryleung-solana** [comment]
CC @t-nelson for input. From the above, it's a little difficult to directly measure memory usage from x509 verification and apparently there are decent ways to rate-limit this in newer versions anyway. I also understand that it was requested to track various state transitions in the handshake, but this doesn't look super feasible.

File: quinn/src/endpoint.rs
Comment (ryleung-solana) [reply quinn/src/endpoint.rs:333]: Don't really have a strong opinion on this. I just thought it was something we might as well include in the stats interface - maybe eventually superseding [Endpoint::open_connections](https://docs.rs/quinn-proto/latest/quinn_proto/struct.Endpoint.html#method.open_connections) though you will know better than me whether this is appropriate.

File: quinn/src/endpoint.rs
Comment (djc) [inline quinn/src/endpoint.rs:347]: If `incoming_handshakes` is merely the sum of accepted, refused and ignored handshakes I think we should track those three numbers separately without additionally tracking `incoming_handshake`, since it can trivially derived from the other three (that is, replace this with `accepted_handshakes`).

File: quinn/src/endpoint.rs
Comment (djc) [inline quinn/src/endpoint.rs:341]: This should be an actual docstring (`///`) without the trailing empty lines.

There should also be a docstring for each of the public fields, in particular taking care to document which of these are monotonically increasing vs other behavior (perhaps we can reflect the difference in the field names, too).

Comment (djc) [reply quinn/src/endpoint.rs]: "Struct for holding the stats returned from Endpoint::stats" is a very mechanical description. Maybe something more like "Statistics on [`Endpoint`] activity"?

Also, note that I requested you document each field here.

**Review by djc — APPROVED** [review]
LGTM, thanks.

**Comment by ryleung-solana** [comment]
Hmm, I'm unable to merge. Perhaps it's because "@Ralith requested changes"?

**Comment by djc** [comment]
You won't be able to merge either way, only we do. But, I'll wait for @Ralith's approval.

**Review by Ralith — APPROVED** [review]
Thanks!

**Pull request merged**
